File: rust/tests/led/sk6812rgbw_test/src/main.rs
use linux_embedded_hal::SpidevBus;
use periph::chips::led::{Sk6812RgbwMinimal, Sk6812RgbwFull};

macro_rules! check_true {
    ($cond:expr, $label:expr, $p:expr, $f:expr) => {
        if $cond { println!("PASS {}", $label); $p += 1; }
        else      { println!("FAIL {}", $label); $f += 1; }
    };
}

macro_rules! check_eq {
    ($got:expr, $exp:expr, $label:expr, $p:expr, $f:expr) => {
        if $got == $exp { println!("PASS {}", $label); $p += 1; }
        else { println!("FAIL {}: got {:?}, expected {:?}", $label, $got, $exp); $f += 1; }
    };
}

fn main() {
    let spi_bus: u8    = std::env::var("SPI_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(0);
    let spi_device: u8 = std::env::var("SPI_DEVICE").ok().and_then(|v| v.parse().ok()).unwrap_or(0);

    let mut passed = 0i32;
    let mut failed = 0i32;

    // --- Sk6812RgbwMinimal ---
    {
        let spi = SpidevBus::open(format!("/dev/spidev{}.{}", spi_bus, spi_device))
            .expect("open spidev");
        let mut strip = Sk6812RgbwMinimal::new(spi, 8);

        check_true!(strip.fill(255, 0, 0, 0).is_ok(), "fill(255,0,0,0) accepted", passed, failed);
        check_true!(strip.fill(0, 255, 0, 0).is_ok(), "fill(0,255,0,0) accepted", passed, failed);
        check_true!(strip.fill(0, 0, 255, 0).is_ok(), "fill(0,0,255,0) accepted", passed, failed);
        check_true!(strip.fill(0, 0, 0, 255).is_ok(), "fill(w=255) accepted", passed, failed);
        check_true!(strip.off().is_ok(), "off() accepted", passed, failed);
    }

    // --- Sk6812RgbwFull ---
    {
        let spi = SpidevBus::open(format!("/dev/spidev{}.{}", spi_bus, spi_device))
            .expect("open spidev");
        let mut strip = Sk6812RgbwFull::new(spi, 8);

        check_eq!(strip.get_brightness(), 255u8, "default brightness is 255", passed, failed);

        strip.set_pixel(0, 255, 0, 0, 0);
        check_true!(strip.show().is_ok(), "set_pixel + show accepted", passed, failed);

        strip.set_pixel(7, 0, 0, 0, 255);
        check_true!(strip.show().is_ok(), "set_pixel(w=255) + show accepted", passed, failed);

        strip.set_brightness(128);
        check_eq!(strip.get_brightness(), 128u8, "brightness setter", passed, failed);
        check_true!(strip.show().is_ok(), "show() with brightness=128 accepted", passed, failed);

        strip.set_brightness(0);
        check_true!(strip.show().is_ok(), "show() with brightness=0 accepted", passed, failed);

        strip.set_brightness(255);
        strip.rotate(1);
        check_true!(strip.show().is_ok(), "rotate + show accepted", passed, failed);

        check_true!(strip.fill_hsv(0.0, 1.0, 1.0).is_ok(), "fill_hsv(0.0) accepted", passed, failed);
        check_true!(strip.fill_hsv(0.333, 1.0, 1.0).is_ok(), "fill_hsv(0.333) accepted", passed, failed);
        check_true!(strip.fill_hsv(0.667, 1.0, 1.0).is_ok(), "fill_hsv(0.667) accepted", passed, failed);
        check_true!(strip.off().is_ok(), "off() on Full accepted", passed, failed);
    }

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/led/sk6812rgbw_test/main.go
//go:build linux && !tinygo

// SK6812RGBW hardware test — Linux host.
//
// Opens /dev/spidevB.D and runs the SK6812RGBW check sequence:
// smoke tests for both Minimal and Full, fill (including the
// dedicated white channel), set_pixel, show, brightness scaling,
// rotation, and HSV fill. Prints PASS/FAIL per check and ends
// with the standard ===DONE: ... === line. Exits 0 on full pass,
// 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/led"
	"github.com/tuhde/Periph/go/periph/connection"
)

func main() {
	bus, err := strconv.Atoi(envOr("SPI_BUS", "0"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "SPI_BUS:", err)
		os.Exit(2)
	}
	dev, err := strconv.Atoi(envOr("SPI_DEVICE", "0"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "SPI_DEVICE:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	conn, trErr := connection.NewNeoPixelConnection(bus, dev, nil)
	if trErr != nil {
		fmt.Fprintln(os.Stderr, "connection open failed (no hw?):", trErr)
		check("sk6812rgbw_minimal_skip_no_hw", true)
		check("sk6812rgbw_full_skip_no_hw", true)
		fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
		if failed != 0 {
			os.Exit(1)
		}
		return
	}
	defer conn.Close()

	// --- SK6812RGBWMinimal ---
	{
		strip, err := led.NewSK6812RGBWMinimal(conn, 8) // Create SK6812RGBW driver, (connection, n=8) → (*SK6812RGBWMinimal, error)
		check("minimal_construct", err == nil && strip != nil)
		if err == nil {
			check("fill_red_ok", strip.Fill(255, 0, 0, 0) == nil)   // Fill red, (r=255, g=0, b=0, w=0) → error
			check("fill_green_ok", strip.Fill(0, 255, 0, 0) == nil) // Fill green, (r=0, g=255, b=0, w=0) → error
			check("fill_blue_ok", strip.Fill(0, 0, 255, 0) == nil)  // Fill blue, (r=0, g=0, b=255, w=0) → error
			check("fill_white_ok", strip.Fill(0, 0, 0, 255) == nil) // Fill white channel, (r=0, g=0, b=0, w=255) → error
			check("off_ok", strip.Off() == nil)                     // Turn off, () → error
		}
	}

	// --- SK6812RGBWFull ---
	{
		strip, err := led.NewSK6812RGBWFull(conn, 8) // Create SK6812RGBW driver, (connection, n=8) → (*SK6812RGBWFull, error)
		check("full_construct", err == nil && strip != nil)
		if err == nil {
			check("default_brightness_255", strip.GetBrightness() == 255) // Get global brightness, () → uint8

			strip.SetPixel(0, 255, 0, 0, 0)              // Set pixel 0 red, (index=0, r=255, g=0, b=0, w=0) → ()
			check("set_pixel_show_ok", strip.Show() == nil) // Transmit buffer, () → error

			strip.SetPixel(7, 0, 0, 0, 255)                       // Set last pixel warm white, (index=7, r=0, g=0, b=0, w=255) → ()
			check("set_pixel_white_show_ok", strip.Show() == nil) // Transmit buffer, () → error

			strip.SetBrightness(128)                                          // Set global brightness, (value=128) → ()
			check("brightness_set_128", strip.GetBrightness() == 128)          // Get global brightness, () → uint8
			check("show_brightness_128_ok", strip.Show() == nil)               // Transmit buffer, () → error

			strip.SetBrightness(0)                                            // Set global brightness, (value=0) → ()
			check("show_brightness_0_ok", strip.Show() == nil)                // Transmit buffer, () → error

			strip.SetBrightness(255)                                  // Set global brightness, (value=255) → ()
			strip.Rotate(1)                                           // Shift buffer left by 1 pixel, (steps=1) → ()
			check("rotate_show_ok", strip.Show() == nil)              // Transmit buffer, () → error

			check("fill_hsv_0_ok", strip.FillHSV(0.0, 1.0, 1.0) == nil)     // Fill red via HSV, (h=0.0, s=1.0, v=1.0) → error
			check("fill_hsv_0333_ok", strip.FillHSV(0.333, 1.0, 1.0) == nil) // Fill green via HSV, (h=0.333, s=1.0, v=1.0) → error
			check("fill_hsv_0667_ok", strip.FillHSV(0.667, 1.0, 1.0) == nil) // Fill blue via HSV, (h=0.667, s=1.0, v=1.0) → error
			check("off_full_ok", strip.Off() == nil)                          // Turn off, () → error
		}
	}

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
